scala-conductr-bundle-lib/src/main/scala/com/typesafe/conductr/bundlelib/scala/LocationCache.scala

```diff
@@ -39,24 +39,23 @@ class LocationCache extends CacheLike {
   val reaperTimer = new Timer()
 
   override def getOrElseUpdate(serviceName: String)(op: => Future[Option[(JavaURI, Option[FiniteDuration])]]): Future[Option[JavaURI]] =
-    cache.getOrElseUpdate(serviceName, {
-      import Implicits.global
-      op
-        .andThen {
-          case Success(Some((_, Some(maxAge)))) =>
+    cache.get(serviceName) match {
+      case None =>
+        import Implicits.global
+        op.flatMap {
+          case Some((uri, Some(maxAge))) =>
             reaperTimer.schedule(new TimerTask {
-              override def run(): Unit = {
+              override def run(): Unit =
                 cache.remove(serviceName)
-              }
             }, maxAge.toMillis)
+            val result = Future.successful(Some(uri))
+            cache.update(serviceName, result)
+            result
           case _ =>
-            cache.remove(serviceName)
+            Future.successful(None)
         }
-        .map {
-          case Some((location, _)) => Some(location)
-          case None                => None
-        }
-    })
+      case Some(v) => v
+    }
```

Comment:
I don't think you've suggested on https://github.com/typesafehub/conductr-lib/pull/107#discussion_r70190863 will work since we'll have a Future containing null value instead.